src/clj/puppetlabs/puppetserver/shell_utils.clj

```diff
@@ -0,0 +1,96 @@
+(ns puppetlabs.puppetserver.shell-utils
+  (:require [schema.core :as schema]
+            [clojure.java.io :as io])
+  (:import (com.puppetlabs.puppetserver ProcessWrapper ExecutionStubImpl)
+           (org.apache.commons.exec CommandLine DefaultExecutor PumpStreamHandler)
+           (java.io IOException ByteArrayInputStream ByteArrayOutputStream)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def ExecutionResult
+  "A map that contains the details of the result of executing a command."
+  {:exit-code schema/Int
+   :stderr schema/Str
+   :stdout schema/Str})
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Internal
+
+(schema/defn ^:private ^:always-validate
+  validate-command
+  "Checks the command string to ensure that is is an absolute path, executable
+  and that the file exists. An exception is thrown is any of those are not the
+  case."
+  [command :- schema/Str]
+  (let [command-file (io/as-file command)]
+    (cond
+      (not (.isAbsolute command-file))
+      (throw (IllegalArgumentException.
+              (format "An absolute path is required, but %s is not an absolute path" command)))
+      (not (.exists command-file))
+      (throw (IllegalArgumentException.
+              (format "The referenced command '%s' does not exist" command)))
+      (not (.canExecute command-file))
+      (throw (IllegalArgumentException.
+              (format "The referenced command '%s' is not executable" command))))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(schema/defn ^:always-validate
+  execute-command :- ExecutionResult
+  "Execute the specified fully qualified command and any included options and
+  return the exit-code, and the contents of stdout and stderr for the command."
```

Comment:
Are the `options` for the command itself, or for this function?  I wonder because there's a general pattern where the last argument of a function can be a map that allows for all sorts of various "options" without having to create an explicit function signature for each optional argument.

So, we should probably just mention that the options will be given to the command and aren't going to be used by this function itself (which was initially what I was expecting).